### DISCUSSION TITLE
Is it possible to make understroke transparent in renders, where `render_bg=False`?

### DISCUSSION THREAD

**Discussion by Mark2Mark** [coldtype/coldtype#130] [General]
Because of this pale color issue, I tried to get away with rendering the frames of an animation into a transparent-bakckground image sequence. However, this animation uses the `understroke()` method and needs a color for that. I’d like to have that color the same as the BG color in the end, but in this approach, that seems not possible? I get black (or whatever I defined as color there) outlines, but wish to just "cut" these out of the overlapping letters.

Is it somehow possible?

**Comment by stenson** [comment]
You mean like you're applying a background color using another program/compositor? Can definitely "knock out" a color a few different ways, either in coldtype itself or in another program.

For coldtype-itself, you can use the `luma` function from `coldtype.fx.skia`:

```python
from coldtype import *
from coldtype.fx.skia import luma

# invert applied colors, then hit with luma

# first approach: add stroke-first (final argument) to fssw
#   > args = (f)ill-(s)troke-(s)troke(w)idth [(strokefirst)]

# this is significantly faster than understroke b/c
# no complex path operations are being done in python

@animation((1080, 300))
def stroke_first_luma(f):
    return (StSt("COLD", Font.ColdObvi(), 300, tu=-100, ro=1)
        .align(f.a.r)
        .fssw(1, 0, 18, 1)
        .f(1)
        .reverse()
        .ch(luma(f.a.r)))

@animation((1080, 300))
def understroke_luma(f):
    return (StSt("COLD", Font.ColdObvi(), 300, tu=-100, ro=1)
        .align(f.a.r)
        .f(1)
        .reverse()
        .understroke(s=0, sw=18)
        .ch(luma(f.a.r)))
```

The caveat here is that you want pure white and black in order to get a good result from `luma`. If you're looking to do color in coldtype, you could do something like this (which is a little more involved, but only a few lines of code):

```python
from coldtype import *
from coldtype.fx.skia import precompose, luma, fill

@renderable((1080, 300))
def stroke_first_luma_with_color(r):
    return (StSt("COLD", Font.ColdObvi(), 300, tu=-100, ro=1)
        .align(r)
        .fssw(1, 0, 18, 1)
        .f(1)
        .reverse()
        .ch(luma(r))
        .ch(precompose(r))
        .ch(fill(hsl(0.3))))
```

That kind of functionality is generally available in video editors as well, sometimes called Luma though sometimes under different names, like Composite Mode or Blend Mode (and then a value like "Screen" might be get a good result).

_Alternate_ idea, which is a little swiss-army-knifey, is to solve the issue in vector-space (rather than image-space as above), and use some boolean path operations to literally cut parts of out adjacent letters, like this (added some extra post-cut letter-spacing to make it obvious that the cuts are "real" and not just composited color):

![transparent_understroke_understroke_cut](https://user-images.githubusercontent.com/89770/218274721-2d1af91f-ed2b-49ac-9435-1337047e7841.png)

```python
from coldtype import *
from functools import partial

@renderable((1080, 680))
def understroke_cut(r):
    def cut(line, i, p):
        if i > 0:
            return p.difference(line[i-1].copy().outline(10, drawOuter=False))

    return (StSt("COLD\nTYPE", Font.ColdObvi(), 300, tu=-100, ro=1)
        .f(hsl(0.9))
        .map(lambda p: p.map(partial(cut, p)))
        .map(lambda p: p.track(50))
        .align(r))
```

**Reply by Mark2Mark** [reply]
Hi Rob,

Thanks so much for all those solutions. Yeah, I tried to get away with blend modes in the video editor, but before I know how to properly knock out, I used a black outline on a white-with-transparency fill and that looked weird with any blend mode. So those luma solutions are already great. Also I didn’t know that Luma is similar to Blend Modes. I tried some things with CT’s Blendmode, but didn’t get far.

The black-and-white only is a little limiting in my case, so that’s where your last example comes into play. That’s super awesome. Thanks for all these ideas!

Also: didn’t know that stroke-first trick. Will definitely replace the understroke in 90% of my animations with it 👍

What do the `precompose` and `partial` do?
